assigner/server/server_test.go

```diff
@@ -0,0 +1,294 @@
+		err = os.Chdir("storetheindex")
+		require.NoError(t, err)
+		e.run("go", "install")
+	*/
+
+	err = os.Chdir(cwd)
+	require.NoError(t, err)
+
+	indexer := filepath.Join(e.dir, "storetheindex")
+	e.run(indexer, "init", "--pubsub-topic", pubsubTopic, "--no-bootstrap", "--block-policy",
+		"--listen-admin=/ip4/127.0.0.1/tcp/3602",
+		"--listen-finder=/ip4/127.0.0.1/tcp/3600",
+		"--listen-ingest=/ip4/127.0.0.1/tcp/3601",
+		"--listen-p2p=/ip4/127.0.0.1/tcp/3603",
+	)
+	stiCfg, err := sticfg.Load(filepath.Join(e.dir, ".storetheindex", "config"))
+	require.NoError(t, err)
+	indexerID := stiCfg.Identity.PeerID
+	t.Log("Initialized indexer", indexerID)
+
+	cmdIndexer := e.start(indexer, "daemon")
+	select {
+	case <-e.indexerReady:
+	case <-ctx.Done():
+		t.Fatal("timed out waiting for indexer to start")
+	}
+
+	rng := rand.New(rand.NewSource(1413))
+
+	// Initialize everything
+	peerID, _, err := pubIdent.Decode()
+	if err != nil {
+		t.Fatal(err)
+	}
+	assigner, cfg := initAssigner(t, pubIdent.PeerID)
+	s := setupServer(t, assigner)
+	cl := setupClient(t, s.URL())
+
+	errChan := make(chan error, 1)
+	go func() {
+		err := s.Start()
+		if err != http.ErrServerClosed {
+			errChan <- err
+		}
+		close(errChan)
+	}()
+
+	ai, err := peer.AddrInfoFromString(fmt.Sprintf("/ip4/127.0.0.1/tcp/9999/p2p/%s", peerID))
+	if err != nil {
+		t.Fatal(err)
+	}
+	ai.ID = peerID
+
+	mhs := util.RandomMultihashes(1, rng)
+
+	assignChan := assigner.OnAssignment(peerID)
+
+	if err := cl.Announce(context.Background(), ai, cid.NewCidV1(22, mhs[0])); err != nil {
+		t.Fatalf("Failed to announce to %s: %s", s.URL(), err)
+	}
+
+	select {
+	case adminURL := <-assignChan:
+		if adminURL != cfg.IndexerPool[0].AdminURL {
+			t.Fatalf("assigned to wrong admin url, expected %s got %s", cfg.IndexerPool[0].AdminURL, adminURL)
+		}
+		t.Log("Assigned publisher to indexer at", adminURL)
+	case <-time.After(5 * time.Second):
+		t.Fatal("timed out waiting for assignment")
+	}
+
+	// Check assignment
+	outProvider := e.run(indexer, "admin", "allow-list", "--indexer", cfg.IndexerPool[0].AdminURL)
+	expect := peerID.String()
+	if !strings.Contains(string(outProvider), expect) {
+		t.Errorf("expected provider to contains %q, got %q", expect, string(outProvider))
+	}
+
+	e.stop(cmdIndexer, 5*time.Second)
+}
+
+// initAssigner initializes a new registry
+func initAssigner(t *testing.T, trustedID string) (*core.Assigner, config.Assignment) {
+	const indexerIP = "127.0.0.1"
+	var cfg = config.Assignment{
+		IndexerPool: []config.Indexer{
+			{
+				AdminURL:  fmt.Sprintf("http://%s:3602", indexerIP),
+				IngestURL: fmt.Sprintf("http://%s:3601", indexerIP),
+			},
+		},
+		Policy: config.Policy{
+			Allow:  false,
+			Except: []string{trustedID},
+		},
+		PubSubTopic: pubsubTopic,
+	}
+	assigner, err := core.NewAssigner(context.Background(), cfg, nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	time.Sleep(5 * time.Second)
```

Comment:
Can we avoid this using `require.Eventually`?

Reply:
I don't think this is needed at all. It was copied from the e2e test.